File: solutions/fortran/reverse-string/1/reverse_string.f90
module reverse_string
  implicit none
contains

  function reverse(input) result(reversed)
    character(*), intent(in) :: input
    character(len=len(input)) :: reversed

    integer :: i, l

    l = len_trim(input) + 1

    do i = 1, l - 1
        reversed(l-i:l-i) = input(i:i)
    end do
  end function
end module

File: solutions/c/reverse-string/1/reverse_string.c
#include "reverse_string.h"
#include <string.h>
#include <stdlib.h>

char *reverse(const char *value) {
    int len = strlen(value);
    char *reversed = malloc(len + 1); // result is freed by caller
    reversed[len] = '\0';

    char *t = reversed + len - 1;
    for (const char *s = value; *s; ++s, --t) {
        *t = *s;
    }
    
    return reversed;
}